core/unit_test/view/TestViewConstructibilityTraits.cpp

```diff
@@ -0,0 +1,85 @@
+//
+//@HEADER
+
+#include <Kokkos_Core.hpp>
+#include <type_traits>
+
+namespace {
+// Relevant documentation:
+// https://kokkos.org/kokkos-core-wiki/API/core/view/view.html#assignment-rules
+// Not everything will be tested here as this is just testing the
+// `std::is_constructible_v` trait
+
+// Differing rank not allowed
+static_assert(
+    !std::is_constructible_v<Kokkos::View<double **>, Kokkos::View<double *>>);
+static_assert(
+    !std::is_assignable_v<Kokkos::View<double **> &, Kokkos::View<double *>>);
+
+// Differing non const value type
+static_assert(
+    !std::is_constructible_v<Kokkos::View<double *>, Kokkos::View<int *>>);
+static_assert(
+    !std::is_assignable_v<Kokkos::View<double *> &, Kokkos::View<int *>>);
+
+// Don't construct away constness
+static_assert(!std::is_constructible_v<Kokkos::View<double *>,
+                                       Kokkos::View<const double *>>);
+static_assert(!std::is_assignable_v<Kokkos::View<double *> &,
+                                    Kokkos::View<const double *>>);
+
+// Construct to const is ok
+static_assert(std::is_constructible_v<Kokkos::View<const double *>,
+                                      Kokkos::View<double *>>);
+static_assert(std::is_assignable_v<Kokkos::View<const double *> &,
+                                   Kokkos::View<double *>>);
+
+// Differing static extents
+static_assert(
+    !std::is_constructible_v<Kokkos::View<double[3]>, Kokkos::View<double[4]>>);
+static_assert(
+    !std::is_assignable_v<Kokkos::View<double[3]> &, Kokkos::View<double[4]>>);
+
+// Differing layouts:
+// Single rank is OK
+static_assert(
+    std::is_constructible_v<Kokkos::View<double[3], Kokkos::LayoutLeft>,
+                            Kokkos::View<double[3], Kokkos::LayoutRight>>);
+static_assert(
+    std::is_constructible_v<Kokkos::View<double[3], Kokkos::LayoutRight>,
+                            Kokkos::View<double[3], Kokkos::LayoutLeft>>);
+static_assert(
+    std::is_assignable_v<Kokkos::View<double[3], Kokkos::LayoutLeft> &,
+                         Kokkos::View<double[3], Kokkos::LayoutRight>>);
+static_assert(
+    std::is_assignable_v<Kokkos::View<double[3], Kokkos::LayoutRight> &,
+                         Kokkos::View<double[3], Kokkos::LayoutLeft>>);
+
+// Rank > 1 not ok
+static_assert(
+    !std::is_constructible_v<Kokkos::View<double[3][5], Kokkos::LayoutRight>,
+                             Kokkos::View<double[3][5], Kokkos::LayoutLeft>>);
+static_assert(
+    !std::is_constructible_v<Kokkos::View<double[3][5], Kokkos::LayoutLeft>,
+                             Kokkos::View<double[3][5], Kokkos::LayoutRight>>);
+static_assert(
+    !std::is_assignable_v<Kokkos::View<double[3][5], Kokkos::LayoutRight> &,
+                          Kokkos::View<double[3][5], Kokkos::LayoutLeft>>);
+static_assert(
+    !std::is_assignable_v<Kokkos::View<double[3][5], Kokkos::LayoutLeft> &,
+                          Kokkos::View<double[3][5], Kokkos::LayoutRight>>);
```

Comment:
Considering how we always check assignable && constructible consider defining a `is_constructible_and_assigable_v` and a `not_constructive_nor_assigable_v` variable templates.